### PR TITLE
feat: azure devops pipeline playbook action

### PR DESCRIPTION
**Pull request by adityathebe**
resolves: https://github.com/flanksource/mission-control/issues/733

### CODE REVIEW

File: api/v1/playbook_actions.go [hunk 1]
```diff
@@ -98,6 +98,26 @@ type GithubAction struct {
...
+
+type AzureDevopsAction struct {
+	// Org is the name of the Azure DevOps organization
+	Org string `json:"org" template:"true"`
+	// Project ID or project name
+	Project string `json:"project" template:"true"`
+	// Token is the personal access token
+	Token types.EnvVar `yaml:"token" json:"token"`
+	// Pipelines is the list of azure pipelines to invoke
+	Pipelines []AzureDevopsPipeline `json:"pipelines,omitempty" template:"true"`
```
Comment (moshloop) [inline api/v1/playbook_actions.go:118]: I don't think we would execute multiple pipelines in 1 action, 1 action == 1 pipeline run so we can inline them here:

The action would take an id for pipeline id, branch and a map of parameters.

File: api/v1/playbook_actions.go [hunk 2]
```diff
@@ -362,6 +382,7 @@ type PlaybookAction struct {
 	Exec         *ExecAction         `json:"exec,omitempty" yaml:"exec,omitempty" template:"true"`
 	GitOps       *GitOpsAction       `json:"gitops,omitempty" yaml:"gitops,omitempty" template:"true"`
 	Github       *GithubAction       `json:"github,omitempty" yaml:"github,omitempty" template:"true"`
+	AzureDevops  *AzureDevopsAction  `json:"azureDevops,omitempty" yaml:"azureDevops,omitempty" template:"true"`
```
Comment (moshloop) [inline api/v1/playbook_actions.go:385]: This should probably be azureDevopsPipeline

**Comment by adityathebe** [comment]
image

File: api/v1/playbook_actions.go [hunk 3]
```diff
@@ -98,6 +98,54 @@ type GithubAction struct {
...
+
+	StagesToSkip []string `json:"stagesToSkip,omitempty" template:"true"`
+
+	// If true, don't actually create a new run.
+	// Instead, return the final YAML document after parsing templates.
+	PreviewRun bool `json:"previewRun,omitempty" template:"true"`
+
+	// If you use the preview run option, you may optionally supply different YAML.
+	// This allows you to preview the final YAML document without committing a changed file.
+	YamlOverride string `json:"yamlOverride,omitempty" template:"true"`
```
Comment (moshloop) [inline api/v1/playbook_actions.go:133]: I don't think using these options would make sense.

File: api/v1/playbook_actions.go [hunk 4]
```diff
@@ -98,6 +98,54 @@ type GithubAction struct {
 	Workflows []GithubWorkflow `yaml:"workflows,omitempty" json:"workflows,omitempty" template:"true"`
 }
 
+type AzureDevopsPipeline struct {
+	// ID is the pipeline ID
+	ID string `json:"id" template:"true"`
+	// Version is the pipeline version
+	Version string `json:"version,omitempty" template:"true"`
+	// ApiVersion is the version of the API to use. Defaults to "7.1-preview.1"
+	ApiVersion string `json:"apiVersion,omitempty" template:"true"`
```
Comment (moshloop) [inline api/v1/playbook_actions.go:107]:
```suggestion
```

**Pull request merged**